src/main/scala/firrtl/passes/Checks.scala

```diff
@@ -155,47 +156,84 @@ object CheckHighForm extends Pass {
       }
       e foreach checkHighFormW(info, mname)
       e foreach checkHighFormT(info, mname)
-      e foreach checkHighFormE(info, mname, names)
+      e foreach checkHighFormE(info, mname, names, literals)
     }
 
     def checkName(info: Info, mname: String, names: NameSet)(name: String): Unit = {
       if (names(name))
         errors.append(new NotUniqueException(info, mname, name))
       names += name
     }
+    
+    def resolveToLiteral(literals: LiteralSet, e: Expression): Boolean = {
+      if(e.isInstanceOf[Literal])
+        return true
+      e match {
+        case t: HasName => literals.contains(t.name) && literals(t.name)
+        case t: WRef => literals.contains(t.name) && literals(t.name)
+        case t: WSubIndex => resolveToLiteral(literals, t.expr)
+        case t: WSubField => resolveToLiteral(literals, t.expr)
+        case t: WSubAccess => resolveToLiteral(literals, t.expr)
+        case _ => false
+      }
+    }
+    
+    def validateLiteral(info: Info, mname: String, literals: LiteralSet, loc: Expression, value: Expression): Unit = {
+      loc match {
+        case t: HasName => validateLiteralS(info, mname, literals, t.name, value)
+        case t: WRef => validateLiteralS(info, mname, literals, t.name, value)
+        case t: WSubIndex => validateLiteral(info, mname, literals, t.expr, value)
+        case t: WSubField => validateLiteral(info, mname, literals, t.expr, value)
+        case t: WSubAccess => validateLiteral(info, mname, literals, t.expr, value)
+        case _ => // do nothing 
+      }
+    }
+    
+    def validateLiteralS(info: Info, mname: String, literals: LiteralSet, name: String, value: Expression): Unit = {
+      if (literals.contains(name))
+        literals(name) = literals(name) && resolveToLiteral(literals, value)
+      else 
+        literals(name) = resolveToLiteral(literals, value)
```

Comment:
Normal Scala style says that multi-line `if-else` should use curly braces.

Reply:
This really doesn't matter, but if you want to have fun with it, you could express this as:
```scala
literals(name) = literals.getOrElse(name, true) && resolveToLiteral(literals, value)
```

Reply:
Nice one-liner indeed